specs/Picker-API_spec.md

```diff
@@ -0,0 +1,121 @@
+   experience currently is to select a single folder and repeat the dialog for more folders. This 
+   is a less than ideal experience for application developers.
+
+**3. Unable to specify an arbitrary start location for the picker.**
+
+   The existing File/Folder pickers do not allow the developer to specify an arbitrary start location (even if the app has access). Currently, the start location is limited to a set of predefined enum values.
+
+The Reunion picker dialog addresses the above issues. This Reunion API will be based upon the below
+APIs in the Windows SDK. The Reunion Picker API will supercede the existing APIs and add additional methods
+for the functionality: File and folder picker, Multiple folder picker.
+
+- Windows.Storage.Pickers.FileOpenPicker (https://docs.microsoft.com/en-us/uwp/api/Windows.Storage.Pickers.FileOpenPicker)
+- Windows.Storage.Pickers.FileSavePicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker)
+- Windows.Storage.Pickers.FolderPicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.folderpicker) 
+
+# Description
+
+The goal of this API is to address the gaps in the existing File/Folder picker APIs. The API surface is similar to the existing
+Picker APIs in Windows SDK with additional methods to support File + Folder picker and multiple folder picker. The Project Reunion APIs will supercede the Windows APIs.
+
+# Examples
+## Show file and folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.FiltersAndDescriptions.Insert("Images", "*.jpg, *.png");
+    picker.StartLocation = startLocation;
+    picker.PickerKinds = StorageItemPickerKinds.Folder | StorageItemPickerKinds.FileOpen;
+    // multi-select will be automatically set
+    IReadOnlyList<Windows.Storage.IStorageItem> storageItems = await picker.ShowPickerAsync();
+
+    if(0 < storageItems.Count)
+    {
+    }
+```
+## Show single folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.Folder;
+    IReadOnlyList<Windows.Storage.IStorageItem> folder = await picker.ShowPickerAsync();
+```
+## Show multi folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.Folder;
+    picker.MultiSelect = true;
+    IReadOnlyList<Windows.Storage.IStorageItem> folders = await picker.ShowPickerAsync();
+```
+## Show multiple file picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.StartLocation = startLocation;
+    picker.MultuSelect = true;
+    picker.PickerKinds = StorageItemPickerKinds.FileOpen;
+    IReadOnlyList<Windows.Storage.IStorageItem> files = await picker.ShowPickerAsync();
+```
+## Show File Save picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.FileSave;
+    IReadOnlyList<Windows.Storage.IStorageItem> fileToSave = await picker.ShowPickerAsync();
```

Comment:
Looks like there's an assumption that this is happening on a UX thread, with the `Show-` verb-part?

Reply:
That is correct. The API must be called on the UI thread. My plan for v0.5 is to use the existing Win32 file dialog until a XAML based picker UI is available. Since we need to the CoreWindow/HWND of the caller, this api must be called on the UI thread.

---
In reply to: [499727076](https://github.com/microsoft/ProjectReunion/pull/99#discussion_r499727076) [](ancestors = 499727076)